spec/support/test_data_cache.rb

```diff
@@ -0,0 +1,62 @@
+# rubocop:disable Style/ClassVars
+class TestDataCache
+  #  A helper class for pre-creating test data with FactoryBot factories
+  #  and maintaining a cache of these objects to avoid the performance problems
+  #  of repeatedly asking FactoryBot to create trees of objects.
+  #
+  #  Refs: https://evilmartians.com/chronicles/testprof-2-factory-therapy-for-your-ruby-tests-rspec-minitest
+  #
+  # The test data and can be accessed as follows:
+  #
+  # TestDataCache.get(:course, :primary, :unpublished)
+  #
+  # Note that a course with the requested traits needs to be pre-defined in this class.
+  # It does not matter what order the traits are when passed as arguments.
+
+  class << self
+    def get(name, *traits)
+      case name
+      when :course
+        if @@cache_populated
+          @@courses[traits.sort] || raise_trait_error(:course, traits)
+        else
+          # fall back to factory (while we migrate all our tests)
+          FactoryBot.create(name, *traits)
```

Comment:
We should be used the factory as defined in `@@course_factories` so we have a single source of truth for how to create this cached data.

```suggestion
          @@course_factories[traits.sort].call
```

Reply:
👍